### PR TITLE
[12.0][MIG] account_operating_unit

### PR DESCRIPTION
**Pull request by bjeficent**
Migration of module to version 12.0

### CODE REVIEW

File: account_operating_unit/models/company.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    inter_ou_clearing_account_id = fields.Many2one('account.account',
+                                                   'Inter-operating unit\
+                                                   clearing account')
+
+    ou_is_self_balanced = fields.Boolean('Operating Units are self-balanced',
+                                         help="Activate if your company is "
+                                              "required to generate a balanced"
+                                              " balance sheet for each "
+                                              "operating unit.",
+                                         default=False)
```
Comment (nikul-serpentcs) [inline account_operating_unit/models/company.py]: @bjeficent Remove `default=False`, `Boolean` is always `False`

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Improve Code

**Comment by bjeficent** [comment]
Updates done. Thanks for reviewing

File: account_operating_unit/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    "data": [
+        "security/account_security.xml",
+        "views/account_move_view.xml",
+        "views/account_journal_view.xml",
+        "views/company_view.xml",
+        "views/invoice_view.xml",
+        "views/account_payment_view.xml",
+        "views/account_invoice_report_view.xml",
+    ],
+    'installable': True,
```
Comment (elicoidal) [inline account_operating_unit/__manifest__.py]: not necessary
add a cover image for apps-store

File: account_operating_unit/models/account_journal.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                                             "journal is used.")
+
+    @api.multi
+    @api.constrains('type')
+    def _check_ou(self):
+        for journal in self:
+            if journal.type in ('bank', 'cash') \
+                    and journal.company_id.ou_is_self_balanced \
+                    and not journal.operating_unit_id:
+                raise UserError(_('Configuration error\nThe operating unit '
```
Comment (elicoidal) [inline account_operating_unit/models/account_journal.py]: Suggestion: "If defined as self-balanced at company level, the operating unit is mandatory in bank journal"

File: account_operating_unit/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    @api.multi
+    @api.constrains('line_ids')
+    def _check_ou(self):
+        for move in self:
+            if not move.company_id.ou_is_self_balanced:
+                continue
+            for line in move.line_ids:
+                if not line.operating_unit_id:
+                    raise UserError(_('Configuration error!\nThe operating\
+                    unit must be completed for each line if the operating\
```
Comment (elicoidal) [inline account_operating_unit/models/account_move.py]:
```suggestion
                    unit is mandatory for each line as the operating\
```

File: account_operating_unit/models/company.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    inter_ou_clearing_account_id = fields.Many2one('account.account',
+                                                   'Inter-operating unit\
+                                                   clearing account')
+
+    ou_is_self_balanced = fields.Boolean('Operating Units are self-balanced',
+                                         help="Activate if your company is "
```
Comment (elicoidal) [inline account_operating_unit/models/company.py]: remove trailing space and add in next line consistently

File: account_operating_unit/models/company.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+                                              " balance sheet for each "
+                                              "operating unit.")
+
+    @api.multi
+    @api.constrains('ou_is_self_balanced')
+    def _inter_ou_clearing_acc_required(self):
+        for rec in self:
+            if rec.ou_is_self_balanced and not \
+                    rec.inter_ou_clearing_account_id:
+                raise UserError(_('Configuration error!\nPlease indicate an\
```
Comment (elicoidal) [inline account_operating_unit/models/company.py]: indicate -> Provide (everywhere)

File: account_operating_unit/readme/CONFIGURE.rst [hunk 7]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (elicoidal) [inline account_operating_unit/readme/CONFIGURE.rst]: remove all starting empty lines

File: account_operating_unit/readme/CONFIGURE.rst [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+If your company is required to generate a balanced balance sheet by
+Operating Unit you can specify at company level that Operating Units should
+be self-balanced, and then indicate a self-balancing clearing account.
+
+1. Create an account "Inter-OU Clearing". It is a balance sheet account.
```
Comment (elicoidal) [inline account_operating_unit/readme/CONFIGURE.rst]:
```suggestion
#. Create an account "Inter-OU Clearing". It is a balance sheet account.
```

File: account_operating_unit/readme/CONFIGURE.rst [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
```
Comment (elicoidal) [inline account_operating_unit/readme/CONFIGURE.rst]: remove empty lines in list

File: account_operating_unit/readme/CONFIGURE.rst [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+2. Go to *Settings / Companies / Configuration* and Set the "Operating Units
```
Comment (elicoidal) [inline account_operating_unit/readme/CONFIGURE.rst]:
```suggestion
#. Go to *Settings / Companies / Configuration* and Set the "Operating Units
```

File: account_operating_unit/readme/CONFIGURE.rst [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+   are self-balanced" checkbox.
+
+   Then set the "Inter-OU Clearing"  account in "Inter-Operating Unit
+   clearing account" field.
+
+3. Go to *Accounting / Configuration / Accounting / Journals* and define, for
```
Comment (elicoidal) [inline account_operating_unit/readme/CONFIGURE.rst]:
```suggestion
#. Go to *Accounting / Configuration / Accounting / Journals* and define, for
```

File: account_operating_unit/readme/DESCRIPTION.rst [hunk 12]
```diff
@@ -0,0 +1,29 @@
+
+This module allows a company to manage the accounting based on Operating
+Units (OU's).
+
+* The financial reports (Trial Balance, P&L, Balance Sheet), allow to report
+  the balances of one or more OU's.
+
```
Comment (elicoidal) [inline account_operating_unit/readme/DESCRIPTION.rst]: remove empty lines

File: account_operating_unit/security/account_security.xml [hunk 13]
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo noupdate="1">
+
+    <record id="ir_rule_account_journal_allowed_operating_units"
+            model="ir.rule">
+        <field name="model_id" ref="account.model_account_journal"/>
+        <field name="domain_force">['|', ('operating_unit_id','=',False), ('operating_unit_id','in',[g.id for g in user.operating_unit_ids])]</field>
```
Comment (elicoidal) [inline account_operating_unit/security/account_security.xml]: improve readability of the line (multiple)

File: account_operating_unit/views/invoice_view.xml [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        <field name="inherit_id" ref="account.invoice_supplier_form" />
+        <field name="arch" type="xml">
+            <field name="journal_id" position="before">
+                <field name="operating_unit_id"
+                       options="{'no_create': True}"
+                       widget="selection" groups="operating_unit.group_multi_operating_unit"/>
+            </field>
+            <xpath  expr="//field[@name='invoice_line_ids']"
+                    position="attributes">
+                <attribute name="context">{'type': type, 'journal_id': journal_id, 'operating_unit': operating_unit_id}</attribute>
```
Comment (elicoidal) [inline account_operating_unit/views/invoice_view.xml]: improve readability

File: account_operating_unit/views/invoice_view.xml [hunk 15]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (elicoidal) [inline account_operating_unit/views/invoice_view.xml:5]: copyright in xml files

**Review by elicoidal — CHANGES_REQUESTED** [review]
mostly nitpicking. no test.

File: account_operating_unit/models/company.py (hunk 5 above)
Comment (nikul-serpentcs) [reply account_operating_unit/models/company.py]: @bjeficent Look like this, Same as all fields.
```
ou_is_self_balanced = fields.Boolean(
    'Operating Units are self-balanced',
    help="""Activate if your company is required to generate a balanced
    balance sheet for each operating unit."""
)
```

File: account_operating_unit/models/invoice.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        return True
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    operating_unit_id = fields.Many2one('operating.unit',
+                                        related='invoice_id.operating_unit_id',
+                                        string='Operating Unit', store=True,
+                                        readonly=True)
```
Comment (nikul-serpentcs) [inline account_operating_unit/models/invoice.py]: @bjeficent `related fields`,  no need to `readonly=True`

File: account_operating_unit/models/account_payment.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    def _compute_operating_unit_id(self):
+        for payment in self:
+            if payment.journal_id:
+                payment.operating_unit_id = \
+                    payment.journal_id.operating_unit_id
+
+    operating_unit_id = fields.Many2one(
+        'operating.unit', string='Operating Unit',
+        domain="[('user_ids', '=', uid)]",
+        compute='_compute_operating_unit_id', readonly=True, store=True)
```
Comment (nikul-serpentcs) [inline account_operating_unit/models/account_payment.py]: @bjeficent remove `readonly=True,` `compute` always readonly

File: account_operating_unit/views/invoice_view.xml [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+        <field name="name">account.invoice.form</field>
+        <field name="model">account.invoice</field>
+        <field name="inherit_id" ref="account.invoice_form" />
+        <field name="arch" type="xml">
+            <field name="journal_id" position="before">
+                <field name="operating_unit_id"
+                       options="{'no_create': True}"
+                       widget="selection" groups="operating_unit.group_multi_operating_unit"/>
+            </field>
+            <xpath  expr="//field[@name='invoice_line_ids']"
```
Comment (nikul-serpentcs) [inline account_operating_unit/views/invoice_view.xml]: @bjeficent use direct `fields`, same as other if possible

**Review by nikul-serpentcs — COMMENTED** [review]
Improve code

File: account_operating_unit/__manifest__.py (hunk 2 above)
Comment (bjeficent) [reply account_operating_unit/__manifest__.py]: As you have said, it is not necessary yet. I could do it later.

**Comment by bjeficent** [comment]
Comments attended. Thanks for reviewing @elicoidal @nikul-serpentcs

File: account_operating_unit/__manifest__.py (hunk 2 above)
Comment (elicoidal) [reply account_operating_unit/__manifest__.py]: Not sure about your interpretation: key is not necessary, image is highly recommended :smile:

**Review by elicoidal — APPROVED** [review]
What about the copyright date. shouldnot it be updated to 2019?

File: account_operating_unit/__manifest__.py [hunk 19]
```diff
@@ -0,0 +1,27 @@
+# © 2016-17 Eficent Business and IT Consulting Services S.L.
+# © 2016 Serpent Consulting Services Pvt. Ltd.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
+{
+    "name": 'Accounting with Operating Units',
+    "summary": "Introduces Operating Unit fields in invoices and "
```
Comment (elicoidal) [inline account_operating_unit/__manifest__.py]: For the sake of consistency with the other modules.
```suggestion
    "summary": "Introduces Operating Unit (OU) in invoices and "
```

File: account_operating_unit/models/account_move.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+                                        'Default operating unit',
+                                        domain="[('user_ids', '=', uid)]",
+                                        help="This operating unit will "
+                                             "be defaulted in the move lines.")
+
+    @api.multi
+    def _prepare_inter_ou_balancing_move_line(self, move, ou_id,
+                                              ou_balances):
+        if not move.company_id.inter_ou_clearing_account_id:
+            raise UserError(_('Error. You need to define an inter-operating\
```
Comment (elicoidal) [inline account_operating_unit/models/account_move.py]: Configuration error

File: account_operating_unit/readme/DESCRIPTION.rst [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+  self-balanced, and the corresponding Inter-Operating Unit clearing account.
+  The Chart of Accounts will always be balanced, for each Operating Unit.
+* A company considering Operating Unit as applicable to report only profits
+  and losses will not need to set the OU's as self-balanced.
+* The self-balancing of Operating Unit is ensured at the time of posting a
+  journal entry. In case that the journal involves posting of items in
+  separate Operating Units, new journal items will be created, using the
+  Inter-Operating Unit clearing account, to ensure that each OU is going to
+  be self-balanced for that journal entry.
+* Adds the Operating Unit (OU) to the invoice. A user can choose what OU to
```
Comment (elicoidal) [inline account_operating_unit/readme/DESCRIPTION.rst]: no need to repeat the definition of OU everywhere in the text. 
Add it at the beginning (introduction for example) and remove everywhere else.

**Review by elicoidal — APPROVED** [review]
LGTM

File: account_operating_unit/__manifest__.py (hunk 2 above)
Comment (bjeficent) [reply account_operating_unit/__manifest__.py]: Oh sorry, I misunderstood you. I have updated that and I agree with you in the image issue, but I don't have any "good" image in hand and neither time to do it right now, sorry.

**Comment by bjeficent** [comment]
Minor changes and copyright updated. Thanks again :smile:

**Comment by bjeficent** [comment]
Ping @nikul-serpentcs 
Could you update your review please? I've updated the code :smile:

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1:

**Review by HviorForgeFlow — CHANGES_REQUESTED** [review]
Functional review:
I follow up instructions:
- Create a Inter-OU clearing account type equity
- Add account to operation unit account on company settings
When I try to add an operating unit to Bank Journal I received this error:
image

Then, I tried to create a Invoice with an operating unit, By default I have Main Operatin Unit as OU on Other info. If I change to another OU like B2C and try to save...:
image

**Comment by bjeficent** [comment]
Comments attended, thanks for the review.
Added #157 fix.

**Comment by AaronHForgeFlow** [comment]
Assigning my user to OU B2B, trying to create a new journal for B2B. I got access error. I restarted the server and try to create the journal again. Then it works  :confused:

**Comment by AaronHForgeFlow** [comment]
After #158 Issues commented here are resolved.

**Comment by AaronHForgeFlow** [comment]
@hveficent could you update your review?

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: LGTM functionally speaking

**Comment by AaronHForgeFlow** [comment]
I tested again and did not get any security issues.

**Review by CasVissers-360ERP — APPROVED** [review]
LGTM, only thing I would change is invoice.py -> account.invoice.py and company.py -> res.company.py

**Comment by meigallodixital** [comment]
LGTM. I revise the latest patched errors and it works fine. Something else to try?

**Review by meigallodixital — APPROVED** [review]
Functional review. LGTM.

**Review by HviorForgeFlow — APPROVED** [review]

File: account_operating_unit/tests/test_account_operating_unit.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+                'account_id': self.env['account.account'].search(
+                    [('user_type_id', '=', acc_type.id)], limit=1).id
+            }
+            lines.append((0, 0, line_values))
+        inv_vals = {
+            'partner_id': self.partner1.id,
+            'account_id': self.partner1.property_account_payable_id.id,
+            'operating_unit_id': operating_unit_id,
+            'name': "Test Supplier Invoice",
+            'reference_type': "none",
```
Comment (nikul-serpentcs) [inline account_operating_unit/tests/test_account_operating_unit.py:153]: @bjeficent @AaronHForgeFlow @HviorForgeFlow  No need to `reference_type` in v12

Comment (AaronHForgeFlow) [reply account_operating_unit/tests/test_account_operating_unit.py:153]: Thanks for the notice. Could you create a PR with the fix?

**Pull request merged**